File: tools/hitechcity/campus_allowlist.js
/**
 * Shared campus-company allowlist for Hitech City board browse
 * (Naukri / Foundit / Cutshort / Instahyre when HITECHCITY_COMPANY_ALLOWLIST is set).
 *
 * Env:
 *   HITECHCITY_COMPANY_ALLOWLIST  path to JSON array of company name strings
 *                                 OR path to companies.json ({ companies: [{name}] })
 *   When unset → allow all (normal portal daily behavior).
 */
"use strict";

const fs = require("fs");
const path = require("path");

let _cache = undefined;

function _norm(s) {
  return String(s || "")
    .toLowerCase()
    .replace(/j\.?\s*p\.?/g, "jp")
    .replace(/[^a-z0-9]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function _nameMatch(target, found) {
  const t = _norm(target);
  const f = _norm(found);
  if (!t || !f) return false;
  if (t.includes(f) || f.includes(t)) return true;
  const tc = t.replace(/\s/g, "");
  const fc = f.replace(/\s/g, "");
  if (tc && (tc.includes(fc) || fc.includes(tc))) return true;
  const tTok = new Set(t.split(" "));
  const fTok = new Set(f.split(" "));
  // JPMorgan / JP Morgan / Chase aliases
  const jpLeft = new Set(["jpmorgan", "jp", "chase", "jpmc", "jpmorganchase", "morgan"]);
  if ([...tTok].some((x) => jpLeft.has(x)) && [...fTok].some((x) => jpLeft.has(x))) return true;
  const overlap = [...tTok].filter((x) => fTok.has(x)).length;
  return overlap >= Math.min(2, tTok.size);
}

function _loadNames() {
  if (_cache !== undefined) return _cache;
  const p = process.env.HITECHCITY_COMPANY_ALLOWLIST;
  if (!p) {
    _cache = null;
    return _cache;
  }
  try {
    const raw = JSON.parse(fs.readFileSync(p, "utf8"));
    let names = [];
    if (Array.isArray(raw)) {
      names = raw.map((x) => (typeof x === "string" ? x : x?.name)).filter(Boolean);
    } else if (Array.isArray(raw.companies)) {
      names = raw.companies.map((c) => c?.name).filter(Boolean);
    } else if (Array.isArray(raw.names)) {
      names = raw.names.filter(Boolean);
    }
    _cache = names.filter(Boolean);
  } catch (e) {
    console.error("[hitechcity-allowlist] failed to load", p, e.message || e);
    _cache = null;
  }
  return _cache;
}

function companyAllowed(companyName) {
  const list = _loadNames();
  if (!list) return true;
  if (!list.length) return false;
  for (const target of list) {
    if (_nameMatch(target, companyName || "")) return true;
  }
  return false;
}

function allowlistActive() {
  return _loadNames() !== null;
}

function resetAllowlistCache() {
  _cache = undefined;
}

module.exports = {
  companyAllowed,
  allowlistActive,
  resetAllowlistCache,
  _norm,
};

File: tools/hitechcity/campus_allowlist.py
"""Campus company allowlist helpers (Python side — Indeed + discovery)."""

from __future__ import annotations

import json
import os
from pathlib import Path

from tools.hitechcity.filters import company_name_match

_COMPANIES_PATH = Path(__file__).with_name("companies.json")
_UNSET = object()
_cache: list[str] | None | object = _UNSET


def allowlist_path() -> str | None:
    return os.environ.get("HITECHCITY_COMPANY_ALLOWLIST") or None


def load_allowlist_names() -> list[str] | None:
    """Return original company names when allowlist env is set; None = no filter."""
    global _cache
    if _cache is not _UNSET:
        return _cache  # type: ignore[return-value]
    p = allowlist_path()
    if not p:
        _cache = None
        return None
    try:
        raw = json.loads(Path(p).read_text(encoding="utf-8"))
        names: list[str] = []
        if isinstance(raw, list):
            names = [x if isinstance(x, str) else (x or {}).get("name") for x in raw]
        elif isinstance(raw, dict):
            if isinstance(raw.get("companies"), list):
                names = [c.get("name") for c in raw["companies"] if isinstance(c, dict)]
            elif isinstance(raw.get("names"), list):
                names = list(raw["names"])
        _cache = [n for n in names if n]
    except Exception:
        _cache = None
    return _cache  # type: ignore[return-value]


def reset_allowlist_cache() -> None:
    global _cache
    _cache = _UNSET


def company_allowed(company_name: str) -> bool:
    names = load_allowlist_names()
    if names is None:
        return True
    if not names:
        return False
    for target in names:
        if company_name_match(target, company_name or ""):
            return True
    return False


def _skip_uhg_enabled() -> bool:
    raw = (os.environ.get("HITECHCITY_SKIP_UHG") or "1").strip().lower()
    return raw not in ("0", "false", "no", "off")


_UHG_ALLOWLIST_SKIP = {
    "optum",
    "unitedhealth group",
    "united health",
    "unitedhealth",
    "uhg",
}


def write_allowlist_artifact(companies: list[dict], dest: Path | None = None) -> Path:
    """Write name-only allowlist JSON for board subprocesses.

    When HITECHCITY_SKIP_UHG is on (default), drop Optum / UnitedHealth Group so
    Naukri/Foundit/etc. do not burn the board cap on Taleo login walls.
    """
    out = dest or Path(
        os.environ.get(
            "HITECHCITY_ALLOWLIST_OUT",
            "/opt/cursor/artifacts/hitechcity-company-allowlist.json",
        )
    )
    if not out.parent.is_dir():
        out = Path(__file__).resolve().parents[2] / "artifacts" / "hitechcity-company-allowlist.json"
    names: set[str] = set()
    for c in companies:
        name = (c.get("name") or "").strip()
        if not name:
            continue
        if _skip_uhg_enabled() and name.lower() in _UHG_ALLOWLIST_SKIP:
            continue
        names.add(name)
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_text(json.dumps({"names": sorted(names)}, indent=2), encoding="utf-8")
    return out


def load_companies_file(path: Path | None = None) -> dict:
    p = path or _COMPANIES_PATH
    return json.loads(p.read_text(encoding="utf-8"))
